File: src/frievald/cpp/algorithms.cpp
#include "algorithms.hpp"
#include <cmath>
#include <algorithm>
#include <iostream>

using namespace std;

// --- Helper Functions ---

// Add two matrices
Matrix add(const Matrix& A, const Matrix& B) {
    size_t n = A.rows();
    size_t m = A.cols();
    Matrix C(n, m);
    for (size_t i = 0; i < n * m; ++i) {
        C.data[i] = A.data[i] + B.data[i];
    }
    return C;
}

// Subtract two matrices
Matrix sub(const Matrix& A, const Matrix& B) {
    size_t n = A.rows();
    size_t m = A.cols();
    Matrix C(n, m);
    for (size_t i = 0; i < n * m; ++i) {
        C.data[i] = A.data[i] - B.data[i];
    }
    return C;
}

// Get next power of 2
size_t next_power_of_2(size_t n) {
    size_t p = 1;
    while (p < n) p <<= 1;
    return p;
}

// Pad matrix to size new_n x new_n
Matrix pad_matrix(const Matrix& A, size_t new_n) {
    size_t n = A.rows();
    Matrix P(new_n, new_n); // Initialized to 0
    for (size_t i = 0; i < n; ++i) {
        for (size_t j = 0; j < n; ++j) {
            P(i, j) = A(i, j);
        }
    }
    return P;
}

// Unpad matrix to original size
Matrix unpad_matrix(const Matrix& P, size_t n) {
    Matrix A(n, n);
    for (size_t i = 0; i < n; ++i) {
        for (size_t j = 0; j < n; ++j) {
            A(i, j) = P(i, j);
        }
    }
    return A;
}

// --- Triple Loop ---

Matrix matmul_triple_loop(const Matrix& A, const Matrix& B) {
    size_t n = A.rows();
    size_t m = A.cols();
    size_t p = B.cols();
    
    // Assuming A is n x m and B is m x p
    // Result is n x p
    Matrix C(n, p);

    // Simple cache-friendly optimization: transpose B? 
    // For strict "triple loop" definition we usually just do ijk.
    // But to be "considerably faster" than Python, even naive C++ is enough.
    // Let's stick to standard ijk for correctness and simplicity first.
    
    for (size_t i = 0; i < n; ++i) {
        for (size_t k = 0; k < m; ++k) {
            double r = A(i, k);
            for (size_t j = 0; j < p; ++j) {
                C(i, j) += r * B(k, j);
            }
        }
    }
    return C;
}

// --- Strassen ---

Matrix strassen_recursive(const Matrix& A, const Matrix& B, int threshold) {
    size_t n = A.rows();

    if (n <= (size_t)threshold) {
        return matmul_triple_loop(A, B);
    }

    size_t k = n / 2;

    // Submatrices
    Matrix A11(k, k), A12(k, k), A21(k, k), A22(k, k);
    Matrix B11(k, k), B12(k, k), B21(k, k), B22(k, k);

    for (size_t i = 0; i < k; ++i) {
        for (size_t j = 0; j < k; ++j) {
            A11(i, j) = A(i, j);
            A12(i, j) = A(i, j + k);
            A21(i, j) = A(i + k, j);
            A22(i, j) = A(i + k, j + k);

            B11(i, j) = B(i, j);
            B12(i, j) = B(i, j + k);
            B21(i, j) = B(i + k, j);
            B22(i, j) = B(i + k, j + k);
        }
    }

    // 7 Products
    // M1 = (A11 + A22)(B11 + B22)
    Matrix M1 = strassen_recursive(add(A11, A22), add(B11, B22), threshold);
    // M2 = (A21 + A22)B11
    Matrix M2 = strassen_recursive(add(A21, A22), B11, threshold);
    // M3 = A11(B12 - B22)
    Matrix M3 = strassen_recursive(A11, sub(B12, B22), threshold);
    // M4 = A22(B21 - B11)
    Matrix M4 = strassen_recursive(A22, sub(B21, B11), threshold);
    // M5 = (A11 + A12)B22
    Matrix M5 = strassen_recursive(add(A11, A12), B22, threshold);
    // M6 = (A21 - A11)(B11 + B12)
    Matrix M6 = strassen_recursive(sub(A21, A11), add(B11, B12), threshold);
    // M7 = (A12 - A22)(B21 + B22)
    Matrix M7 = strassen_recursive(sub(A12, A22), add(B21, B22), threshold);

    // Reconstruct C
    // C11 = M1 + M4 - M5 + M7
    Matrix C11 = add(sub(add(M1, M4), M5), M7);
    // C12 = M3 + M5
    Matrix C12 = add(M3, M5);
    // C21 = M2 + M4
    Matrix C21 = add(M2, M4);
    // C22 = M1 - M2 + M3 + M6
    // Note: The implementation below uses add/sub helpers which create new Matrix objects by value.
    // This is inefficient for large matrices due to allocation overhead.
    // For production code, use in-place operations or a library like BLAS/Eigen.
    Matrix C22 = add(add(sub(M1, M2), M3), M6);

    Matrix C(n, n);
    for (size_t i = 0; i < k; ++i) {
        for (size_t j = 0; j < k; ++j) {
            C(i, j) = C11(i, j);
            C(i, j + k) = C12(i, j);
            C(i + k, j) = C21(i, j);
            C(i + k, j + k) = C22(i, j);
        }
    }

    return C;
}

Matrix matmul_strassen(const Matrix& A, const Matrix& B, int threshold) {
    size_t n = A.rows();
    // Check if square and power of 2
    size_t m = next_power_of_2(n);
    
    if (m == n) {
        return strassen_recursive(A, B, threshold);
    } else {
        Matrix Ap = pad_matrix(A, m);
        Matrix Bp = pad_matrix(B, m);
        Matrix Cp = strassen_recursive(Ap, Bp, threshold);
        return unpad_matrix(Cp, n);
    }
}

// --- Frievald ---

bool frievald_verify(const Matrix& A, const Matrix& B, const Matrix& C, int k, mt19937& rng, bool force_iterations) {
    size_t n = A.rows();
    uniform_int_distribution<int> dist(0, 1); // Random 0 or 1
    bool all_correct = true;

    for (int iter = 0; iter < k; ++iter) {
        // Generate random vector r
        Matrix r(n, 1);
        for (size_t i = 0; i < n; ++i) {
            r(i, 0) = dist(rng);
        }

        // Compute Br = B * r
        Matrix Br = matmul_triple_loop(B, r);
        // Compute ABr = A * (Br)
        Matrix ABr = matmul_triple_loop(A, Br);
        // Compute Cr = C * r
        Matrix Cr = matmul_triple_loop(C, r);

        // Check if ABr == Cr
        for (size_t i = 0; i < n; ++i) {
            if (abs(ABr(i, 0) - Cr(i, 0)) > 1e-9) {
                if (!force_iterations) {
                    return false;
                }
                all_correct = false;
            }
        }
    }
    return all_correct;
}

// --- Error Injection ---

Matrix inject_error(const Matrix& C, const string& mode, mt19937& rng) {
    Matrix C_err = C;
    size_t n = C.rows();
    
    if (mode == "none") {
        return C_err;
    } else if (mode == "random_element") {
        uniform_int_distribution<size_t> dist_idx(0, n - 1);
        size_t r = dist_idx(rng);
        size_t c = dist_idx(rng);
        C_err(r, c) += 10.0; // Significant error
    } else if (mode == "random_row") {
        uniform_int_distribution<size_t> dist_idx(0, n - 1);
        size_t r = dist_idx(rng);
        for (size_t j = 0; j < n; ++j) {
            C_err(r, j) += 5.0;
        }
    } else if (mode == "random_col") {
        uniform_int_distribution<size_t> dist_idx(0, n - 1);
        size_t c = dist_idx(rng);
        for (size_t i = 0; i < n; ++i) {
            C_err(i, c) += 5.0;
        }
    }
    return C_err;
}

File: src/frievald/python/algorithms.py
"""Core algorithms for matrix multiplication and verification.

This module provides both deterministic matrix multiplication baselines and
Frievald's randomized verification algorithm. All routines operate on square
matrices represented as NumPy ndarrays. Frievald's algorithm runs in
O(k * n^2) time, providing a fast probabilistic verifier for A * B = C.
"""

from __future__ import annotations

from typing import Optional

import numpy as np

EPSILON = 1e-8  # tolerance for floating point comparisons


def _validate_square_matrices(a: np.ndarray, b: np.ndarray, c: np.ndarray) -> int:
    """Validate shapes and return the matrix dimension."""

    # Guard against nonsensical inputs before any heavy computation runs.
    if a.ndim != 2 or b.ndim != 2 or c.ndim != 2:
        raise ValueError("All inputs must be 2D matrices.")
    if a.shape[1] != b.shape[0] or a.shape[0] != c.shape[0] or b.shape[1] != c.shape[1]:
        raise ValueError("Matrix dimensions are incompatible for multiplication.")
    if a.shape[0] != a.shape[1]:
        raise ValueError("Frievald's algorithm assumes square matrices.")
    return a.shape[0]


def frievald_verify(
    a: np.ndarray,
    b: np.ndarray,
    c: np.ndarray,
    k: int = 1,
    rng: Optional[np.random.Generator] = None,
    epsilon: float = EPSILON,
) -> bool:
    """Probabilistically verify whether A * B == C using Frievald's algorithm.

    Runs in O(k * n^2) time where k is the number of repetitions and n is the
    matrix dimension. If the product is incorrect, the probability of a false
    positive is at most (1/2)^k.
    """

    n = _validate_square_matrices(a, b, c)
    if k <= 0:
        raise ValueError("k must be positive.")
    rng = rng or np.random.default_rng()

    dtype = np.result_type(a.dtype, b.dtype, c.dtype)
    # Ensure contiguous buffers so repeated matmul calls avoid implicit copies.
    a = np.ascontiguousarray(a, dtype=dtype)
    b = np.ascontiguousarray(b, dtype=dtype)
    c = np.ascontiguousarray(c, dtype=dtype)

    # Scratch buffers reused across iterations to limit allocations.
    tmp = np.empty(n, dtype=dtype)
    left = np.empty(n, dtype=dtype)
    right = np.empty(n, dtype=dtype)

    for _ in range(k):
        # Random {0,1} witness vector defining the projection tested this round.
        r = rng.integers(0, 2, size=n, dtype=np.int8)
        np.matmul(b, r, out=tmp)
        np.matmul(a, tmp, out=left)
        np.matmul(c, r, out=right)
        if not np.allclose(left, right, atol=epsilon):
            return False
    return True


def matmul_numpy(a: np.ndarray, b: np.ndarray) -> np.ndarray:
    """Compute matrix product using NumPy's optimized matmul (O(n^3))."""

    # np.matmul delegates to BLAS when available, so rely on it for the baseline.
    return np.matmul(a, b)


def matmul_triple_loop(a: np.ndarray, b: np.ndarray, block_size: int = 64) -> np.ndarray:
    """Compute matrix product using a cache-friendly blocked triple-loop implementation."""

    if a.shape[1] != b.shape[0]:
        raise ValueError("Matrix dimensions are incompatible for multiplication.")
    if block_size <= 0:
        raise ValueError("block_size must be positive.")

    a_c = np.ascontiguousarray(a)
    b_c = np.ascontiguousarray(b)
    n, m = a_c.shape
    p = b_c.shape[1]

    result = np.zeros((n, p), dtype=a.dtype)
    b_t = np.ascontiguousarray(b_c.T)

    block = max(1, block_size)

    for i0 in range(0, n, block):
        i_end = min(i0 + block, n)
        a_block = a_c[i0:i_end]
        c_block = result[i0:i_end]
        for j0 in range(0, p, block):
            j_end = min(j0 + block, p)
            for k0 in range(0, m, block):
                k_end = min(k0 + block, m)
                # Blocked multiply reuses tiles in cache and leverages vectorized matmul.
                c_block[:, j0:j_end] += a_block[:, k0:k_end] @ b_t[j0:j_end, k0:k_end].T
                # Tight innermost loop stays in NumPy space, so Python overhead is low.
    return result


def _next_power_of_two(n: int) -> int:
    """Return the smallest power of two greater than or equal to ``n``."""
    return 1 if n == 0 else 2 ** (n - 1).bit_length()


def _pad_to_power_of_two(a: np.ndarray, size: int) -> np.ndarray:
    """Pad matrix ``a`` with zeros to match a square matrix of ``size``."""
    if a.shape[0] == size and a.shape[1] == size:
        return a
    padded = np.zeros((size, size), dtype=a.dtype)
    padded[: a.shape[0], : a.shape[1]] = a
    return padded


def matmul_strassen(a: np.ndarray, b: np.ndarray, threshold: int = 64) -> np.ndarray:
    """Multiply two square matrices using Strassen's algorithm.

    Recursively applies Strassen when size > threshold; below threshold uses
    NumPy's matmul. Pads inputs up to next power-of-two dimension, then trims
    result back to original shape.
    Complexity: O(n^{log_2 7}) ≈ O(n^2.807).
    """

    if a.shape[1] != b.shape[0]:
        raise ValueError("Matrix dimensions are incompatible for multiplication.")
    n = a.shape[0]
    m = _next_power_of_two(n)
    if n != m or a.shape[1] != m or b.shape[1] != m:
        # Pad to power-of-two square before recursing so Strassen splits evenly.
        a_pad = np.ascontiguousarray(_pad_to_power_of_two(a, m))
        b_pad = np.ascontiguousarray(_pad_to_power_of_two(b, m))
    else:
        a_pad = np.ascontiguousarray(a)
        b_pad = np.ascontiguousarray(b)

    def _strassen(x: np.ndarray, y: np.ndarray) -> np.ndarray:
        """Recursively multiply two equally-sized square matrices via Strassen."""
        size = x.shape[0]
        if size <= threshold:
            # Small blocks fall back to fast dense multiplication.
            return x @ y
        mid = size // 2
        a11 = x[:mid, :mid]
        a12 = x[:mid, mid:]
        a21 = x[mid:, :mid]
        a22 = x[mid:, mid:]
        b11 = y[:mid, :mid]
        b12 = y[:mid, mid:]
        b21 = y[mid:, :mid]
        b22 = y[mid:, mid:]

        # Strassen's seven products (reduced multiplication count vs naive eight).
        p1 = _strassen(a11 + a22, b11 + b22)
        p2 = _strassen(a21 + a22, b11)
        p3 = _strassen(a11, b12 - b22)
        p4 = _strassen(a22, b21 - b11)
        p5 = _strassen(a11 + a12, b22)
        p6 = _strassen(a21 - a11, b11 + b12)
        p7 = _strassen(a12 - a22, b21 + b22)

        c11 = p1 + p4 - p5 + p7
        c12 = p3 + p5
        c21 = p2 + p4
        c22 = p1 - p2 + p3 + p6

        c = np.empty((size, size), dtype=x.dtype)
        # Reassemble the four quadrants into a single result matrix.
        c[:mid, :mid] = c11
        c[:mid, mid:] = c12
        c[mid:, :mid] = c21
        c[mid:, mid:] = c22
        return c

    c_pad = _strassen(a_pad, b_pad)
    # Remove padding so the returned matrix matches original dimensions.
    return c_pad[: a.shape[0], : b.shape[1]]


def matrices_equal(a: np.ndarray, b: np.ndarray, epsilon: float = EPSILON) -> bool:
    """Return True if matrices are equal within the given tolerance."""

    if a.shape != b.shape:
        return False
    return np.allclose(a, b, atol=epsilon)
